### PR TITLE
Add two records: CreateDirectories and DirectoryExists. Automatically create directories based on this PV like in AD

### PR DESCRIPTION
**Pull request by jwlodek**
Marking as draft until I can write tests and perform a full test with a real PandA system, but this works in my initial testing.

Designed to work just like the AD `CreateDirectory` system: https://github.com/areaDetector/ADCore/blob/631fb284dc06d52b9286e62418a80f7f3a7a05e2/docs/ADCore/NDPluginFile.rst?plain=1#L57-L64

The `DirectoryExists` record is also meant to behave the same as in areaDetector, and prevent attempting to write the HDF file if the target does not exist or is not writable by the user running the IOC process.

### CODE REVIEW

File: src/pandablocks_ioc/_hdf_ioc.py
Comment (AlexanderWells-diamond) [inline src/pandablocks_ioc/_hdf_ioc.py:576]: I think this code can be simplified by making use of `pathlib.Path` and its platform-agnostic handling of directory paths. For example:
```
def _update_full_file_path(self, new_val):
    # This assumes new_val is a full file path relative to
    # the current directory e.g. dir1/dir2/file.hdf
    new_path = Path(new_val)

    val = self._create_directory_record.get()
    if val < 0:
        create_dir_count = abs(val)
    else:
        create_dir_count = len(new_path.parent.parts) - val

    # Iterate from top-most directory downwards, creating directories as
    # necessary up to maximum amount
    for p in reversed(new_path.parents):
        if not p.exists() and abs(create_dir_count) > 0:
            p.mkdir()
            create_dir_count -= 1

    if new_path.parent.is_dir() and os.access(new_path.parent, os.W_OK):
        self._directory_exists_record.set(1)
    else:
        self._directory_exists_record.set(0)
```
I haven't rigorously tested this, as I don't entirely understand the reasoning behind positive values of CreateDirectory. I don't understand why you would ever want to partially create a directory tree, or have such fine-grained control as to whether the directory is created or not, so these semantics are probably not represented correctly in this example code.

Comment (jwlodek) [reply src/pandablocks_ioc/_hdf_ioc.py:576]: I think the reason for not auto-creating a directory tree is because of the expectation that part of the tree is expected to already exist with certain ownership permissions, and having the IOC process touch these directories may lead to permissions/ownership potentially changing in unwanted ways. From my view auto-creating all dirs in the tree is fine, but I think we should have the same behavior as AD for portability sake in things like writers in ophyd.

**Comment by jwlodek** [comment]
Ok, I think this is ready for review. All tests are passing + tests added for new conditions, and I also tested with a real PandA, everything behaves as expected.

File: tests/test_hdf_ioc.py
Comment (AlexanderWells-diamond) [inline tests/test_hdf_ioc.py:420]: I think this runs the risk of potentially creating the `/panda_test5` folder in someone's root directory, if permissions are not as we expect?

File: tests/test_hdf_ioc.py
Comment (AlexanderWells-diamond) [inline tests/test_hdf_ioc.py:418]: A matter of taste, but `Path` objects support using `/` to join two together:
```suggestion
    target_path = str(tmp_path / path)
```

File: tests/test_hdf_ioc.py
Comment (AlexanderWells-diamond) [inline tests/test_hdf_ioc.py:433]: It would be nice to add `timeout=TIMEOUT` to all these `ca*` calls, just to ensure a known maximum termination time. I do note we haven't been particularly good about always doing it in other tests.

File: tests/test_hdf_ioc.py
Comment (AlexanderWells-diamond) [inline tests/test_hdf_ioc.py:909]: This code should be in the `hf5_controller` fixture, not added to every test

File: src/pandablocks_ioc/_hdf_ioc.py
Comment (AlexanderWells-diamond) [inline src/pandablocks_ioc/_hdf_ioc.py:631]: Should this condition have a test explicitly for it? I note there are tests where it must be `1` or they break, but I'd like to see one that does expect the break to occur.

File: tests/test_hdf_ioc.py
Comment (AlexanderWells-diamond) [inline tests/test_hdf_ioc.py:403]: I'm unsure as to the reason behind this comment; we shouldn't care what the final status message is, every test should be resetting it on each iteration and no bleed-through from one test to another should occur?

**Review by AlexanderWells-diamond — CHANGES_REQUESTED** [review]
I have a couple of changes that I think should be made, and then a lot of nitpicking comments.

I also am not the person to give the final approval to this, as I'm somewhat detached from this project these days, so please wait on @evalott100 to give final approval.

File: src/pandablocks_ioc/_hdf_ioc.py
Comment (jwlodek) [reply src/pandablocks_ioc/_hdf_ioc.py:592]: Yes, this was an oversight - I had made everything print to stderr during testing so it wasn't separated in pytest output.

File: src/pandablocks_ioc/_hdf_ioc.py
Comment (jwlodek) [reply src/pandablocks_ioc/_hdf_ioc.py:566]: how about `max_dirs_to_create` and `dirs_to_create`?

File: tests/test_hdf_ioc.py
Comment (jwlodek) [reply tests/test_hdf_ioc.py:420]: Yes, I guess so. Is there a directory I can use as one that is expected to exist but not be owned by the user running the tests?

File: tests/test_hdf_ioc.py
Comment (jwlodek) [reply tests/test_hdf_ioc.py]: This was a leftover from me being confused at a test failure. Turns out it was catching a typo with the status message created by the callback, which was `Ok` instead of `OK`. Removed this.

File: src/pandablocks_ioc/_hdf_ioc.py
Comment (jwlodek) [reply src/pandablocks_ioc/_hdf_ioc.py:633]: Added test to account for this.

File: tests/test_hdf_ioc.py
Comment (jwlodek) [reply tests/test_hdf_ioc.py]: I ended up just changing the permissions for the `tmp_path` dir at the start of the test to something not writable, then changing them back after.

**Review by AlexanderWells-diamond — APPROVED** [review]
This all looks good to me now

**Comment by GDYendell** [comment]
Is the directory depth here absolute rather than relative to how much of the requested path already exists? i.e. if directory `/1/2/3` exists and I want it to create directory `/1/2/3/4/5`, does `CreateDirectory` need to be to 2 or 5? I had assumed it would be 2.

**Comment by jwlodek** [comment]
@evalott100 Could we get this merged and included in a new release? I'd like to make some dependent changes in ophyd-async to add the option to use this feature.

**Review by evalott100 — CHANGES_REQUESTED** [review]
Nice work! This will be really useful.

It'd maybe be good to add docs on this: https://pandablocks.github.io/PandABlocks-ioc/main/how-to/capture-hdf.html

A copy of the areaDetector description of `CreateDirectory` should suffice.

File: src/pandablocks_ioc/_hdf_ioc.py
Comment (evalott100) [inline src/pandablocks_ioc/_hdf_ioc.py:561]: Need a test covering line 561

File: src/pandablocks_ioc/_hdf_ioc.py
Comment (evalott100) [inline src/pandablocks_ioc/_hdf_ioc.py:594]: Either this should be debug like the next log on success, or both should be info.

**Comment by evalott100** [comment]
Run /tests/regenerate_test_bobfiles.sh to regenerate the test screens

File: docs/how-to/capture-hdf.md
Comment (evalott100) [inline docs/how-to/capture-hdf.md:12]:
``` suggestion
- The number of directories that the IOC is allowed to create provided they don't exist is determined by `:DATA:CreateDirectory`. The behavior of this signal is the same as the identical PV in [`areaDetector`](https://areadetector.github.io/areaDetector/ADCore/NDPluginFile.html).
```

File: src/pandablocks_ioc/_hdf_ioc.py
Comment (jwlodek) [reply src/pandablocks_ioc/_hdf_ioc.py]: There was a test that was expecting "OK" as the output when the dir existed, so I had left it like this, but in the most recent commit I just changed the test to reflect the new status MSG.

**Review by evalott100 — APPROVED** [review]
Nice work! :slightly_smiling_face: Lgtm if it's been tested on a real panda

**Pull request merged**
